### ISSUE TITLE
no-fn-reference-in-iterator is not compatible with TypeScript type guards

### ISSUE THREAD

**Issue by moltar** [sindresorhus/eslint-plugin-unicorn#781]
```ts
function isDefined<T>(argument: T | undefined): argument is T {
  return typeof argument !== 'undefined'
}

const items = [1, 2, undefined]

items.filter(isDefined).map((number) => {
  // number is now type guarded and undefined is eliminated
  console.log(number)
})

items.filter((num) => isDefined(num)).map((number) => {
  // however, this way it does not work, and "number" can still be undefined
  console.log(number)
})
```

I am not sure why TypeScript is not inferring the results in the second way.

**Comment by fisker** [comment]
Console shows the same result on [typescript playground](https://www.typescriptlang.org/play/index.html?ssl=4&ssc=2&pln=1&pc=1#code/GYVwdgxgLglg9mABDAzgEQKbBmDATAHgBUA+ACgEMAnAcxAFsMwoAuRIxAH0XDyx3wBKNtTqNmyFO0QBvAFCJEVDFBBUkUAJ4AHDHGCJRDJlEQBCALwXEAcl79ceG3IC+cuRAQpTMKBnpS1gDaAIwANIgATBH22I4Auu6+-igAdNgANn5UZKiYcUKp9BTaZGRgDABGGFSCiBYksgqIAPQtiBX01VSSHXAA7ohauoh01Hx4hmCTsQKTqIgYGTD0OBR+eM2eYChwGRipGXA05VU1gq4XcskB6TBZNWWddQ2S+XOn9IKCRSVPZ7V6o15Io2ogABYDDAANxqESg4IW-QommQpjwcAwUjAcFM-TgVAA1hEKNNEAAiTrdcmICCkxDee4ZRDVHjTBz4LZePYHI4nKnnS5AA)

**Comment by moltar** [comment]
That's not true:

---

`number` is not undefined:

Screenshot:

<img width="654" alt="image" src="https://user-images.githubusercontent.com/491247/85364480-67fd8a80-b54d-11ea-9d07-05549508f54a.png">

---

`number` is: `number | undefined` type.

Screenshot:

<img width="517" alt="image" src="https://user-images.githubusercontent.com/491247/85364494-7055c580-b54d-11ea-9a9d-af90e3ad7f12.png">

---

I think my example was poor, here is a better one:

https://www.typescriptlang.org/play/index.html?target=99#code/GYVwdgxgLglg9mABDAzgEQKbBmDATAHgBUA+ACgEMAnAcxAFsMwoAuRIxAH0XDyx3wBKNtTqNmyFO0QBvAFCJEVDFBBUkUAJ4AHDHGCJRDJlEQBCALwXEAcl79ceG3IC+cuRAQpTMKBnpS1gDaAIwANIgATBH22I4Auu6+-igAdNgANn5UZKiYcUKp9BTaZGRgDABGGFSCiBYksgqIAPQtiBX01VSSHXAA7ohauoh01Hx4hmCTsQKTqIgYGTD0OBR+eM3KquodVTWugkl+AekwWTVlnXUNkvlz5QyCgkUlV-u19Y3yim2IABYDDAANxqESg-wW-QommQpjwcAwUjAcFM-TgVAA1hEKNNEAAiTrdfGICC4xDec4ZRDVHjTBz4LYqNRIIk1RAAakQIUOQA

**Comment by fisker** [comment]
Sorry, I mean complied js code result.

**Comment by moltar** [comment]
But the code being validated is the TypeScript one. So I’m not sure how the result is relevant?

**Comment by fisker** [comment]
It should be 

```ts
items.filter(((num) => isDefined(num)) as typeof isDefined).map((number) => {
  // however, this way it does not work, and "number" can still be undefined
  return number + 1
})
```
?

**Comment by moltar** [comment]
I am not sure what is going on there. I am getting confused.

Both of the code snippets I posted are valid.

My only concern is that the rule is being triggered in a case where there is no other option. The only option is to do `filter(isDefined)`, as that is how TypeScript works/infers.

**Comment by fisker** [comment]
It's not only about `.filter`, all function references lost types.

**Comment by oss6** [comment]
Hi, the issue is described here: https://github.com/microsoft/TypeScript/issues/10734 and here: https://github.com/microsoft/TypeScript/issues/16069.
Would it be sensible to add an exception for functions with type guards (either by default or through an option)?

**Comment by moltar** [comment]
@oss6 thanks for finding the proper references!

**Comment by moltar** [comment]
I think the rule is renamed to `unicorn/no-array-callback-reference` now as I understand.

**Comment by osdiab** [comment]
I agree that this particular scenario is one where array callbacks is actually nice to use. I'd be interested if the `unicorn/no-array-callback-reference` rule was configurable to allow it specifically for `filter` calls, since wrapping a function in TypeScript loses the type guard, making this non-ergonomic

**Comment by Vages** [comment]
I could try my hand at this. My initial plan is to add an option with a name like `ignoreFilterWithTypeGuards` to the rule. If this is set to true, the linter will ignore any imported function considered a type guard.

I'm good with abstract syntax trees, but not very experienced with writing eslint rules. Does anyone know of any existing rules that rely on the Typescript return type of something that's imported? I've tried thinking about this for 15 minutes without coming up with a single one.

**Comment by fisker** [comment]
> Does anyone know of any existing rules that rely on the Typescript return type of something that's imported?

I don't know about this, but you can try check https://github.com/typescript-eslint/typescript-eslint, maybe they do this.

**Comment by robeady** [comment]
I met this problem too. I'd like to keep using the rule but I can see no convenient workaround. Would a more practical solution be to let the user configure the list of allowed function names, either for filter callbacks or any callback? Introducing type sensitive lints seems pretty drastic for such an isolated issue.

**Comment by RebeccaStevens** [comment]
The typeguard info is lost because the wrapper function is not a typeguard. If you simply make your wrapper function a typeguard then problem fixed.

e.g.

```ts
items.filter((num): num is number => isDefined(num)).map((number) => {
  console.log(number + 1);
});
```

[playground link](https://www.typescriptlang.org/play?target=99#code/GYVwdgxgLglg9mABDAzgEQKbBmDATAHgBUA+ACgEMAnAcxAFsMwoAuRIxAH0XDyx3wBKNtTqNmyFO0QBvAFCJEVDFBBUkUAJ4AHDHGCJRDJlEQBCALwXEAcl79ceG3IC+cuRAQpTMKBnpS1gDaAIwANIgATBH22I4AugDc7gD0KYhoMMrQPCgYLHK+-igAdNgANn5UZKiYcUIl9BTaZGRgDABGGFSCiBYksgqInmAocOUYJeVwNG2d3YgA1Ighgskua6npAMIU3ogA7lTNulQFaYgAcnB+bFAAFqiG2tpUcBQQ94j3e4gPGIgUBRGJIUCAAQ8KKZyjhYGAaEoQBNJH8qJocAioHBEK8MAA3EyIAC0gOxrzgHQoHXKmkQXX2YBuf2xeGxD1QJUKfgCZRglW6rTm9F6-UkdQEeCFgl6vy0un0YocDSaLSFXR6fQG8kUIzGEymMzVC2Wq3WmzkF22IG8cHofx0GDo1DwBSKPIqVVa7WFbG9KO96s1ivqku90sazS98w1ou1wy840m01mAeNK02G2SQA)

Though this being said, typeguards probably should be ignored by this rule (probably via an option like @Vages suggests (I would suggest the option be named `ignoreTypeGuards` or `allowTypeGuards`)) as typeguards are highly unlikely to change their parameter signature.

> This should be trivial for cases where the function exists in the same file
> I'm not sure about what to do when the function is imported, though.

It shouldn't make any difference if the function is defined in the same file or not; all the information needed is in the function's type information.

If people want this, I can make a PR (I'm the main contributor for [eslint-plugin-functional](https://www.npmjs.com/package/eslint-plugin-functional) so I'm pretty experienced with working with eslint rules).

**Comment by tetarchus** [comment]
Was there ever any update on this? Currently can work around with disable comment, but would be nice to be able to remove them.

**Comment by briavicenti** [comment]
+1, I'd love to see a fix to this rule so I can stop disabling it on all inline typeguards.
